File: webapp/Locust/Scripts/app/services/StepService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('StepService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/Step/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                SaveArray: {
                    method: 'POST',
                    isArray: true
                },
                DeleteArray: {
                    method: 'POST',
                    isArray: true
                },
                UpdateArray: {
                    method: 'POST',
                    isArray: true
                },
                Update: {
                    method: 'PUT'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetAll: {
                    method: 'GET',
                    isArray: true
                },
                DeleteForTC: {
                    method: 'DELETE'
                },
                GetForTestCase: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestScenario: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestProcedure: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestSuplemental: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestProcedureSTP: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestScenarioSTP: {
                    method: 'GET',
                    isArray: true
                }
            }),

            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };

        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };

        self.Delete = function (id, success, error) {
            return resource.Delete({ action: 'Delete', id: id }, success || noop, error || noop);
        };

        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };

        self.SaveArray = function (entity, projectId, TypeOfSave, success, error) {
            return resource.SaveArray({ action: 'SaveArray', projectId: projectId, TypeOfSave: TypeOfSave }, entity, success || noop, error || noop);
        };


        self.DeleteArray = function (entity, success, error) {
            return resource.DeleteArray({ action: 'DeleteArray' }, entity, success || noop, error || noop);
        };

        self.UpdateArray = function (entity, projectId, TypeOfSave, TestEvidenceId, Evidence, success, error) {
            return resource.UpdateArray({ action: 'UpdateArray', projectId: projectId, TypeOfSave: TypeOfSave, TestEvidenceId: TestEvidenceId, Evidence: Evidence }, entity, success || noop, error || noop);
        };

        self.DeleteForTC = function (tcId, success, error) {
            return resource.DeleteForTC({ action: 'DeleteForTC', tcId: tcId }, success || noop, error || noop);
        };

        self.GetForTestCase = function (tcId, success, error) {
            return resource.GetForTestCase({ action: 'GetForTestCase', tcId: tcId }, success || noop, error || noop);
        };

        self.GetForTestScenario = function (tsId, success, error) {
            return resource.GetForTestCase({ action: 'GetForTestScenario', tsId: tsId }, success || noop, error || noop);
        };

        self.GetForTestProcedure = function (tpId, success, error) {
            return resource.GetForTestProcedure({ action: 'GetForTestProcedure', tpId: tpId }, success || noop, error || noop);
        };

        self.GetForTestProcedureSTP = function (projectId, tpId, success, error) {
            return resource.GetForTestProcedureSTP({ action: 'GetForTestProcedureSTP', projectId: projectId, tpId: tpId }, success || noop, error || noop);
        };

        self.GetForTestScenarioSTP = function (projectId, tpId, success, error) {
            return resource.GetForTestScenarioSTP({ action: 'GetForTestScenarioSTP', projectId: projectId, tpId: tpId }, success || noop, error || noop);
        };

        self.GetForTestSuplemental = function (stpId, success, error) {
            return resource.GetForTestCase({ action: 'GetForTestSuplemental', stpId: stpId }, success || noop, error || noop);
        };

        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };

    }]);
})(angular);

File: webapp/Locus.Core/Services/StepService.cs
using Locus.Core.DTO;
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface IStepService
    {
        Step Save(Step step);

        Step[] SaveArray(Step[] steps, int projectId ,int TypeOfSave);

        Step[] UpdateArray(Step[] steps, int projectId, int TypeOfSave, int TestEvidenceId, int Evidence);

        Step[] DeleteArray(Step[] steps);

        Step Update(Step step);

        Step Delete(int idStep);

        List<Step> GetAll();

        Step Get(int idStep);

        List<Step> DeleteForTC(int tcId);

        List<Step> GetForTestCase(int tcId);

        List<Step> GetForTestScenario(int tsId);

        List<Step> GetForTestProcedure(int tpId);

        List<Step> GetForTestSuplemental(int stpId);

        List<Step> GetForTestCaseOrder(int tcId);

        List<Step> GetForTestScenarioOrder(int tsId);

        List<Step> GetForTestProcedureOrder(int tpId);

        List<Step> GetForTestSuplementalOrder(int stpId);

        List<StepDTO> GetForTestProcedureSTP(int projectId, int tpId);

        List<StepDTO> GetForTestScenarioSTP(int projectId, int tpId);

    }



    public class StepService : IStepService
    {
        //Instance of ITestCaseRepository
        private readonly IStepRepository _stepRepository;

        //Constructor of TestCaseService and initialize testCaseRepository
        public StepService(IStepRepository stepRepository)
        {
            _stepRepository = stepRepository;
        }

        public Step Delete(int idStep)
        {
            return _stepRepository.Delete(idStep);
        }

        public Step[] DeleteArray(Step[] steps)
        {
            return _stepRepository.DeleteArray(steps);
        }

        public List<Step> DeleteForTC(int tcId)
        {
            return _stepRepository.DeleteForTC(tcId);
        }

        public Step Get(int idStep)
        {
            return _stepRepository.Get(idStep);
        }

        public List<Step> GetAll()
        {
            return _stepRepository.GetAll();
        }

        public List<Step> GetForTestCase(int tcId)
        {
            return _stepRepository.GetForTestCase(tcId);
        }

        public List<Step> GetForTestCaseOrder(int tcId)
        {
            return _stepRepository.GetForTestCaseOrder(tcId);
        }

        public List<Step> GetForTestProcedure(int tpId)
        {
            return _stepRepository.GetForTestProcedure(tpId);
        }

        public List<Step> GetForTestProcedureOrder(int tpId)
        {
            return _stepRepository.GetForTestProcedureOrder(tpId);
        }

        public List<StepDTO> GetForTestProcedureSTP(int projectId, int tpId)
        {
            return _stepRepository.GetForTestProcedureSTP(projectId, tpId);
        }

        public List<Step> GetForTestScenario(int tsId)
        {
            return _stepRepository.GetForTestScenario(tsId);
        }

        public List<Step> GetForTestScenarioOrder(int tsId)
        {
            return _stepRepository.GetForTestScenarioOrder(tsId);
        }

        public List<StepDTO> GetForTestScenarioSTP(int projectId, int tpId)
        {
            return _stepRepository.GetForTestScenarioSTP(projectId, tpId);
        }

        public List<Step> GetForTestSuplemental(int stpId)
        {
            return _stepRepository.GetForTestSuplemental(stpId);
        }

        public List<Step> GetForTestSuplementalOrder(int stpId)
        {
            return _stepRepository.GetForTestSuplementalOrder(stpId);
        }

        public Step Save(Step step)
        {
            if (step.number_steps != 0)
                return _stepRepository.Save(step);
            else
            {
                return null;
            }

        }

        public Step[] SaveArray(Step[] steps, int projectId , int TypeOfSave)
        {
            return _stepRepository.SaveArray(steps, projectId ,TypeOfSave);
        }

        public Step Update(Step step)
        {
            return _stepRepository.Update(step);
        }

        public Step[] UpdateArray(Step[] steps, int projectId, int TypeOfSave, int TestEvidenceId, int Evidence)
        {
            return _stepRepository.UpdateArray(steps, projectId, TypeOfSave, TestEvidenceId, Evidence);
        }
    }
}
